cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+
+### Lifecycle / operations walkthrough
+
+#### Connection
+From scratch, the following example assumes that SPIRE has been bootstrapped in the cluster and the spire-agent running as a container within the cilium-agent pod on each node. Additionally, a CiliumNetworkPolicy must be inserted that selects the target pods and enables the authentication options. We assume that these changes have been calculated and applied to the policymap entries for the pod in the example below.
+
+When a pod sends a packet for a new connection that is selected by a CiliumNetworkPolicy which marks that connectivity as requiring authentication, the BPF datapath goes through a series of steps, from evaluating the network policy to checking the Authentication map (auth). When the packet hits the policymap lookup, the map entry will signal that the connection requires authentication. This triggers an additional check in the datapath into the bpf auth map to determine the authentication status between the peers involved in the connection. For the example below, we assume that this initial authentication check fails because the authentication has not yet occurred for these peers.
+
+![SPIFFE and SPIRE Architecture with Cilium Datapath](./images/SPIFFE-SPIRE-arch-with-datapath.png)
+
+Steps following an authentication map lookup failure:
+1. The packet is dropped. The BPF datapath emits a drop event to userspace notifying that the packet between these peers has been dropped due to the reason that "authentication is required" for this connection.
+1. Cilium agent intercepts the event and initiates a request for certificate material that is necessary in order to establish a TLS session with the cilium-agent on the node where the peer application resides.
+1. Cilium agent initiates the mTLS handshake with the peer node
+1. Cilium agent populates the BPF auth map to indicate that these peers have authenticated, and as such subsequent packets between these peers can perform the lookup in this map and pass the policy/auth check and hence continue to the destination.
+
+The above outline describes the process at the source node. A similar process needs to occur at the destination node, to allow the destination node to authenticate the source pod. A naive approach that should work is to execute the exact same logic at the destination node when packets are processed at ingress towards the destination pod. An improved implementation would initiate this process at the destination node when the source node initiates step (3) towards the destination node.
+
+### Debuggability
+* In-Pod CLI, extend cilium bugtool to gather this data so it gets into sysdumps:
+  * cilium status - overview of authentication status
+    * N/M pods configured for authentication
+    * N authentication sessions with peers
+  * cilium auth list - List all authentication sessions in cilium-agent
+    * Local node
+    * Peer node
+    * Local pod (selector?)
```

Comment:
`Local pod` & `Peer pod` -> `Local Identity` & `Peed Identity`?

the idea of using a dedicated `auth` map instead of e.g. `conntrack` is to have this more identity based.